# source: Trawled out of Red Hat's kernel-2.4.21-32.0.1.EL.src.rpm by Horms
# inclusion: upstream code has been reworked and doesn't appear vulerable
# descrition: on 64 bit architectures incorect handling of xattr offsets
#             may cause a local DoS
# revision date: Fri, 29 Jul 2005 12:04:57 +0900

--- kernel-source-2.4.27.orig/fs/ext3/xattr.c	2005-05-19 19:29:38.000000000 +0900
+++ kernel-source-2.4.27/fs/ext3/xattr.c	2005-07-29 12:01:33.000000000 +0900
@@ -636,10 +636,9 @@
 				goto cleanup;
 			memcpy(header, HDR(bh), bh->b_size);
 			header->h_refcount = cpu_to_le32(1);
-			offset = (char *)here - bh->b_data;
-			here = ENTRY((char *)header + offset);
-			offset = (char *)last - bh->b_data;
-			last = ENTRY((char *)header + offset);
+			offset = (char *)header - bh->b_data;
+			here = ENTRY((char *)here + offset);
+			last = ENTRY((char *)last + offset);
 		}
 	} else {
 		/* Allocate a buffer where we construct the new block. */
